File: ios/DawsonsCompanion/DawsonsCompanion/Synth.swift
import AVFoundation
import Foundation

/// Instrument families available on any timeline track. Mirrors the
/// browser DAW's family set (website/js/demo-songs.js) so the mental
/// model stays consistent across platforms.
enum InstrumentFamily: String, CaseIterable, Codable, Identifiable {
    case drums, keys, guitar, bass, lead, pad, brass, bell, flute, saxophone, clarinet
    case strings, organ, epiano, choir, synthbass, marimba, trumpet
    case voice, anySound

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .drums: return "Drums"
        case .keys: return "Piano"
        case .guitar: return "Guitar"
        case .bass: return "Bass"
        case .lead: return "Lead Synth"
        case .pad: return "Pad"
        case .brass: return "Brass"
        case .bell: return "Bell"
        case .flute: return "Flute"
        case .saxophone: return "Saxophone"
        case .clarinet: return "Clarinet"
        case .strings: return "Strings"
        case .organ: return "Organ"
        case .epiano: return "Electric Piano"
        case .choir: return "Choir"
        case .synthbass: return "Synth Bass"
        case .marimba: return "Marimba"
        case .trumpet: return "Trumpet"
        case .voice: return "Voice"
        case .anySound: return "Any Sound"
        }
    }

    /// A vibrant hue per family, matching the web app's night-stage
    /// palette (website/css/daw.css) exactly, so the two DAWs feel like
    /// the same product despite the native UI.
    var colorHex: UInt32 {
        switch self {
        case .drums: return 0xFF9457
        case .keys: return 0xB478FF
        case .guitar: return 0x35E6AE
        case .bass: return 0xFF5FA8
        case .lead: return 0x3FC7FF
        case .pad: return 0xB478FF
        case .brass: return 0xFF9457
        case .bell: return 0x35E6AE
        case .flute: return 0xFF5FA8
        case .saxophone: return 0xFFAB2E
        case .clarinet: return 0x2FE0C0
        case .strings: return 0xC264FF
        case .organ: return 0xC8EA3F
        case .epiano: return 0x39C9FF
        case .choir: return 0xFF5C96
        case .synthbass: return 0x3FC7FF
        case .marimba: return 0xFFAB2E
        case .trumpet: return 0xFF9457
        case .voice: return 0xFF5C96
        case .anySound: return 0xC8EA3F
        }
    }
}

enum DrumSound: String, CaseIterable, Codable, Identifiable {
    case kick, snare, hihat, clap
    var id: String { rawValue }
    var displayName: String { rawValue.capitalized }
}

private enum OscShape {
    case sine, sawtooth, square, triangle
}

private struct VoiceShape {
    let osc: OscShape
    let attack: Double
    let decay: Double
    let sustain: Double
    let release: Double
    let brightness: Double
}

/// Renders instrument notes directly into PCM buffers using simple
/// oscillator synthesis — no bundled soundfonts or sample libraries,
/// the same "generate everything" approach as the browser's synth.js
/// and desktop's synth.rs.
enum Synth {
    static let sampleRate: Double = 44100

    private static let shapes: [InstrumentFamily: VoiceShape] = [
        .keys: VoiceShape(osc: .triangle, attack: 0.005, decay: 0.25, sustain: 0.4, release: 0.3, brightness: 0.6),
        .guitar: VoiceShape(osc: .sawtooth, attack: 0.005, decay: 0.15, sustain: 0.5, release: 0.25, brightness: 0.5),
        .bass: VoiceShape(osc: .sine, attack: 0.01, decay: 0.1, sustain: 0.8, release: 0.2, brightness: 0.3),
        .lead: VoiceShape(osc: .sawtooth, attack: 0.02, decay: 0.1, sustain: 0.7, release: 0.2, brightness: 0.8),
        .pad: VoiceShape(osc: .triangle, attack: 0.4, decay: 0.3, sustain: 0.8, release: 0.8, brightness: 0.4),
        .brass: VoiceShape(osc: .sawtooth, attack: 0.05, decay: 0.1, sustain: 0.7, release: 0.15, brightness: 0.7),
        .bell: VoiceShape(osc: .sine, attack: 0.001, decay: 0.6, sustain: 0.2, release: 0.8, brightness: 0.9),
        .flute: VoiceShape(osc: .sine, attack: 0.08, decay: 0.1, sustain: 0.7, release: 0.2, brightness: 0.5),
        .saxophone: VoiceShape(osc: .sawtooth, attack: 0.06, decay: 0.1, sustain: 0.7, release: 0.2, brightness: 0.6),
        .clarinet: VoiceShape(osc: .square, attack: 0.05, decay: 0.1, sustain: 0.7, release: 0.2, brightness: 0.4),
        // Bowed strings: slow attack and a long ringing release, the
        // opposite envelope shape from a plucked instrument like guitar.
        .strings: VoiceShape(osc: .sawtooth, attack: 0.18, decay: 0.15, sustain: 0.85, release: 0.5, brightness: 0.5),
        // Organ: instant on, full sustain, essentially no decay stage —
        // its signature envelope even without additive drawbar synthesis.
        .organ: VoiceShape(osc: .square, attack: 0.005, decay: 0.01, sustain: 1.0, release: 0.1, brightness: 0.5),
        // Electric piano (Rhodes-style): a short bell-like attack that
        // decays quickly to a much quieter sustain.
        .epiano: VoiceShape(osc: .triangle, attack: 0.005, decay: 0.35, sustain: 0.25, release: 0.3, brightness: 0.6),
        .choir: VoiceShape(osc: .triangle, attack: 0.3, decay: 0.2, sustain: 0.85, release: 0.6, brightness: 0.5),
        // Synth bass: punchier and more clipped than the sine-wave
        // acoustic bass — a square wave with a faster decay.
        .synthbass: VoiceShape(osc: .square, attack: 0.005, decay: 0.15, sustain: 0.5, release: 0.08, brightness: 0.4),
        // Mallet/percussive: near-instant attack straight into a fast
        // decay with almost no sustain — opposite shape from every
        // sustained wind/string voice above.
        .marimba: VoiceShape(osc: .sine, attack: 0.001, decay: 0.35, sustain: 0.05, release: 0.15, brightness: 0.9),
        .trumpet: VoiceShape(osc: .sawtooth, attack: 0.02, decay: 0.08, sustain: 0.75, release: 0.1, brightness: 0.7),
        .voice: VoiceShape(osc: .triangle, attack: 0.02, decay: 0.1, sustain: 0.7, release: 0.25, brightness: 0.55),
    ]

    static func frequency(forMIDI note: Int) -> Double {
        440.0 * pow(2.0, (Double(note) - 69.0) / 12.0)
    }

    /// Renders a single melodic note as a mono PCM buffer.
    static func renderNote(family: InstrumentFamily, midiNote: Int, durationSec: Double, velocity: Double = 0.9) -> AVAudioPCMBuffer {
        let shape = shapes[family] ?? shapes[.keys]!
        let freq = frequency(forMIDI: midiNote)
        let totalSec = durationSec + shape.release
        let frameCount = AVAudioFrameCount(max(1, totalSec * sampleRate))
        let format = AVAudioFormat(standardFormatWithSampleRate: sampleRate, channels: 1)!
        let buffer = AVAudioPCMBuffer(pcmFormat: format, frameCapacity: frameCount)!
        buffer.frameLength = frameCount
        let data = buffer.floatChannelData![0]

        var phase = 0.0
        let phaseInc = freq / sampleRate

        for i in 0..<Int(frameCount) {
            let t = Double(i) / sampleRate
            let env = envelope(t: t, noteDur: durationSec, shape: shape)
            let raw = oscillate(shape: shape.osc, phase: phase, brightness: shape.brightness)
            data[i] = Float(raw * env * velocity)
            phase += phaseInc
            if phase >= 1 { phase -= 1 }
        }
        return buffer
    }

    /// Renders one hit of a drum voice, matching the browser pattern
    /// editor's drum palette (kick/snare/hihat/clap).
    static func renderDrumHit(sound: DrumSound, velocity: Double = 0.9) -> AVAudioPCMBuffer {
        let durationSec: Double
        switch sound {
        case .kick: durationSec = 0.28
        case .snare: durationSec = 0.2
        case .hihat: durationSec = 0.08
        case .clap: durationSec = 0.18
        }
        let frameCount = AVAudioFrameCount(durationSec * sampleRate)
        let format = AVAudioFormat(standardFormatWithSampleRate: sampleRate, channels: 1)!
        let buffer = AVAudioPCMBuffer(pcmFormat: format, frameCapacity: frameCount)!
        buffer.frameLength = frameCount
        let data = buffer.floatChannelData![0]

        var rngState: UInt64 = 0x9E37_79B9_7F4A_7C15
        func nextNoise() -> Double {
            rngState = rngState &* 6364136223846793005 &+ 1
            return (Double(rngState >> 11) / Double(1 << 53)) * 2 - 1
        }

        for i in 0..<Int(frameCount) {
            let t = Double(i) / sampleRate
            let progress = t / durationSec
            var sample = 0.0
            switch sound {
            case .kick:
                let freq = 120.0 * exp(-t * 18)
                sample = sin(2 * .pi * freq * t)
            case .snare:
                sample = nextNoise() * 0.6 + sin(2 * .pi * 180 * t) * 0.4
            case .hihat:
                sample = nextNoise()
            case .clap:
                sample = nextNoise() * (progress < 0.5 ? 1.0 : 0.6)
            }
            let env = pow(1.0 - progress, 2.0)
            data[i] = Float(sample * env * velocity)
        }
        return buffer
    }

    private static func envelope(t: Double, noteDur: Double, shape: VoiceShape) -> Double {
        if t < shape.attack {
            return t / max(shape.attack, 0.0001)
        }
        let sinceAttack = t - shape.attack
        if sinceAttack < shape.decay {
            let d = sinceAttack / max(shape.decay, 0.0001)
            return 1.0 - d * (1.0 - shape.sustain)
        }
        if t < noteDur {
            return shape.sustain
        }
        let sinceRelease = t - noteDur
        if sinceRelease < shape.release {
            return shape.sustain * (1.0 - sinceRelease / shape.release)
        }
        return 0.0
    }

    private static func oscillate(shape: OscShape, phase: Double, brightness: Double) -> Double {
        switch shape {
        case .sine:
            return sin(2 * .pi * phase)
        case .sawtooth:
            let base = 2 * (phase - floor(phase + 0.5))
            return base * (0.6 + brightness * 0.4)
        case .square:
            return (phase < 0.5 ? 1.0 : -1.0) * (0.5 + brightness * 0.3)
        case .triangle:
            return 2 * abs(2 * (phase - floor(phase + 0.5))) - 1
        }
    }
}

File: website/js/synth.js
// Original, from-scratch Web Audio synthesis engine for the browser DAW.
// No SoundFont here (148MB isn't reasonable to ship on a marketing
// page) — instead a small set of instrument "voices" built from basic
// oscillators + envelopes, the same general technique used by countless
// original synthesizers; not derived from any specific product.

function midiToFreq(note) {
  return 440 * Math.pow(2, (note - 69) / 12);
}

// Renders one note/chord event into an existing AudioBuffer at a given
// offset, using a simple oscillator + ADSR-style gain envelope shaped by
// the instrument "family". Additive: multiple pitches (a chord) just
// call this once per pitch.
function renderVoice(ctx, buffer, family, pitches, startSec, durationSec, sampleRate) {
  const shapes = {
    keys: { wave: "triangle", attack: 0.005, decay: 0.15, sustain: 0.5, release: 0.2, gain: 0.22 },
    guitar: { wave: "sawtooth", attack: 0.005, decay: 0.25, sustain: 0.25, release: 0.15, gain: 0.16 },
    bass: { wave: "sine", attack: 0.01, decay: 0.08, sustain: 0.85, release: 0.1, gain: 0.35 },
    lead: { wave: "triangle", attack: 0.02, decay: 0.1, sustain: 0.7, release: 0.15, gain: 0.2 },
    pad: { wave: "sawtooth", attack: 0.25, decay: 0.2, sustain: 0.8, release: 0.4, gain: 0.14 },
    brass: { wave: "sawtooth", attack: 0.03, decay: 0.1, sustain: 0.7, release: 0.12, gain: 0.18 },
    bell: { wave: "sine", attack: 0.002, decay: 0.6, sustain: 0.2, release: 0.4, gain: 0.18 },
    flute: { wave: "sine", attack: 0.04, decay: 0.1, sustain: 0.75, release: 0.15, gain: 0.2 },
    // A sax's real timbre is complex (breath noise, reed buzz), but a
    // sawtooth with a slower attack than guitar/brass and a warmer,
    // longer sustain gives it a reasonably distinct "breathy horn" feel
    // in this simple oscillator-based synth.
    saxophone: { wave: "sawtooth", attack: 0.06, decay: 0.15, sustain: 0.8, release: 0.2, gain: 0.19 },
    clarinet: { wave: "square", attack: 0.05, decay: 0.08, sustain: 0.85, release: 0.12, gain: 0.16 },
    // A bowed string's attack is slow and its tail rings on after the
    // note ends — the opposite envelope shape from a plucked instrument
    // like guitar, which is the actual distinguishing feature here.
    strings: { wave: "sawtooth", attack: 0.18, decay: 0.15, sustain: 0.85, release: 0.5, gain: 0.15 },
    // An organ's real timbre is additive (multiple fixed-ratio drawbars),
    // which this simple oscillator model can't reproduce — but its
    // signature *envelope* (instant on, full sustain, no decay stage at
    // all until release) is just as recognizable and easy to model.
    organ: { wave: "square", attack: 0.002, decay: 0.001, sustain: 1.0, release: 0.08, gain: 0.16 },
    // Electric piano (Rhodes-style): a short bell-like attack transient
    // that decays quickly to a much quieter sustain — quite different
    // from the acoustic `keys` envelope above.
    epiano: { wave: "triangle", attack: 0.002, decay: 0.35, sustain: 0.25, release: 0.3, gain: 0.2 },
    choir: { wave: "triangle", attack: 0.3, decay: 0.2, sustain: 0.85, release: 0.6, gain: 0.13 },
    // A synth bass is punchier and more clipped-sounding than the sine-
    // wave acoustic `bass` above — a square wave with a faster decay.
    synthbass: { wave: "square", attack: 0.005, decay: 0.15, sustain: 0.5, release: 0.08, gain: 0.3 },
    // Mallet/percussive pitched instrument: near-instant attack straight
    // into a fast decay, almost no sustain — the opposite envelope shape
    // from every sustained wind/string instrument above.
    marimba: { wave: "sine", attack: 0.001, decay: 0.35, sustain: 0.05, release: 0.15, gain: 0.24 },
    trumpet: { wave: "sawtooth", attack: 0.02, decay: 0.08, sustain: 0.75, release: 0.1, gain: 0.2 },
  };
  const shape = shapes[family] || shapes.keys;

  for (const pitch of pitches) {
    const freq = midiToFreq(pitch);
    const startSample = Math.floor(startSec * sampleRate);
    const totalSamples = Math.floor(durationSec * sampleRate);
    const data0 = buffer.getChannelData(0);
    const data1 = buffer.numberOfChannels > 1 ? buffer.getChannelData(1) : data0;

    const attackSamples = Math.floor(shape.attack * sampleRate);
    const decaySamples = Math.floor(shape.decay * sampleRate);
    const releaseSamples = Math.floor(shape.release * sampleRate);
    const sustainSamples = Math.max(0, totalSamples - attackSamples - decaySamples - releaseSamples);

    for (let i = 0; i < totalSamples; i++) {
      const t = i / sampleRate;
      let envelope;
      if (i < attackSamples) {
        envelope = i / Math.max(1, attackSamples);
      } else if (i < attackSamples + decaySamples) {
        const p = (i - attackSamples) / Math.max(1, decaySamples);
        envelope = 1.0 - p * (1.0 - shape.sustain);
      } else if (i < attackSamples + decaySamples + sustainSamples) {
        envelope = shape.sustain;
      } else {
        const p = (i - attackSamples - decaySamples - sustainSamples) / Math.max(1, releaseSamples);
        envelope = shape.sustain * Math.max(0, 1.0 - p);
      }

      let sample;
      const phase = 2 * Math.PI * freq * t;
      if (shape.wave === "sine") {
        sample = Math.sin(phase);
      } else if (shape.wave === "triangle") {
        sample = (2 / Math.PI) * Math.asin(Math.sin(phase));
      } else if (shape.wave === "square") {
        // A clarinet's real timbre is dominated by odd harmonics (it
        // behaves acoustically like a closed pipe) — a square wave is
        // the classic, simple synthesis approximation of that same
        // odd-harmonic-only spectrum, a textbook technique.
        sample = Math.sign(Math.sin(phase));
      } else {
        // sawtooth
        const cycles = freq * t;
        sample = 2 * (cycles - Math.floor(cycles + 0.5));
      }

      const value = sample * envelope * shape.gain;
      const idx = startSample + i;
      if (idx >= 0 && idx < data0.length) {
        data0[idx] += value;
        data1[idx] += value;
      }
    }
  }
}

// Renders a simple drum hit (kick/snare/hihat) as noise/pitch-drop
// bursts, since there's no sample library here — good enough for demo
// songs, not a full drum-machine emulation.
function renderDrumHit(buffer, kind, startSec, sampleRate) {
  const data0 = buffer.getChannelData(0);
  const data1 = buffer.numberOfChannels > 1 ? buffer.getChannelData(1) : data0;
  const startSample = Math.floor(startSec * sampleRate);

  if (kind === "kick") {
    const dur = 0.15;
    const n = Math.floor(dur * sampleRate);
    for (let i = 0; i < n; i++) {
      const t = i / sampleRate;
      const freq = 120 * Math.exp(-t * 18);
      const env = Math.exp(-t * 14);
      const value = Math.sin(2 * Math.PI * freq * t) * env * 0.6;
      const idx = startSample + i;
      if (idx >= 0 && idx < data0.length) {
        data0[idx] += value;
        data1[idx] += value;
      }
    }
  } else if (kind === "snare") {
    const dur = 0.12;
    const n = Math.floor(dur * sampleRate);
    for (let i = 0; i < n; i++) {
      const t = i / sampleRate;
      const env = Math.exp(-t * 22);
      const value = (Math.random() * 2 - 1) * env * 0.35;
      const idx = startSample + i;
      if (idx >= 0 && idx < data0.length) {
        data0[idx] += value;
        data1[idx] += value;
      }
    }
  } else if (kind === "hihat") {
    const dur = 0.05;
    const n = Math.floor(dur * sampleRate);
    for (let i = 0; i < n; i++) {
      const t = i / sampleRate;
      const env = Math.exp(-t * 60);
      const value = (Math.random() * 2 - 1) * env * 0.18;
      const idx = startSample + i;
      if (idx >= 0 && idx < data0.length) {
        data0[idx] += value;
        data1[idx] += value;
      }
    }
  } else if (kind === "clap") {
    // A real handclap is several near-simultaneous noise bursts (multiple
    // hands/fingers not landing at exactly the same instant), so this
    // layers three short, fast-decaying noise bursts a few milliseconds
    // apart rather than one single envelope like the snare above.
    const burstOffsetsSec = [0, 0.012, 0.024];
    const dur = 0.09;
    const n = Math.floor(dur * sampleRate);
    for (const offsetSec of burstOffsetsSec) {
      const offsetSamples = Math.floor(offsetSec * sampleRate);
      for (let i = 0; i < n; i++) {
        const t = i / sampleRate;
        const env = Math.exp(-t * 35);
        const value = (Math.random() * 2 - 1) * env * 0.3;
        const idx = startSample + offsetSamples + i;
        if (idx >= 0 && idx < data0.length) {
          data0[idx] += value;
          data1[idx] += value;
        }
      }
    }
  }
}

export { renderVoice, renderDrumHit, midiToFreq };
